### PR TITLE
Lightweight exceptions

### PR DESCRIPTION
**Pull request by mk13**
Created custom exception class to be used instead of standard Exception - goal is to reduce the weight of transferring exceptions. Previous exception was designed was for simple throwing - not accumulation. 

Future considerations:
- Implementing ErrorCode(Analyzer) and adding to every exception.

### CODE REVIEW

**Review by MichaelRFairhurst — APPROVED** [review]
Seems like they should probably use an error type rather than strings, otherwise we won't be able to easily merge them into our own analysis errors.

They could also use offset+length rather than offset+lexeme, though that difference seems not super important to me

**Comment by mk13** [comment]
I had similar thoughts :) This is just the precursor - expect to see another PR within the week with ErrorCodes added into it to make the error types more distinct and also more easily mergeable into angular_analyzer.

I'll also switch over to offset in that same iteration. After doing some analysis - you're correct, there really is no use in holding the actual lexeme in the exception since whatever uses these exceptions have access to the raw text and can simply get the string on demand. This should lead to more lightweight exceptions.

**Comment by matanlurey** [comment]
Let me know when you rebase and I'll review.

**Comment by mk13** [comment]
Hi @matanlurey, I've merged the previous PR#26 and this is now ready for review. Thanks!

File: lib/src/exception_handler/exception_handler.dart [hunk 1]
```diff
@@ -20,10 +21,10 @@ class ThrowingExceptionHandler implements ExceptionHandler {
 }
 
 class RecoveringExceptionHandler implements ExceptionHandler {
-  final List<Exception> exceptions = new List<Exception>();
+  final List<AngularParserException> exceptions = <AngularParserException>[];
```
Comment (liuming0) [inline lib/src/exception_handler/exception_handler.dart]: List<AngularParserException> can be removed.

File: lib/src/scanner.dart [hunk 2]
```diff
@@ -33,8 +34,11 @@ class NgScanner {
     return _current;
   }
 
-  factory NgScanner(String html, ExceptionHandler exceptionHandler,
-      {sourceUrl}) {
+  factory NgScanner(
+    String html,
+    ExceptionHandler exceptionHandler, {
+    sourceUrl,
```
Comment (liuming0) [inline lib/src/scanner.dart]: What is the type of sourceUrl?

**Review by matanlurey — APPROVED** [review]

**Pull request merged**